### ISSUE TITLE
Make UnsafeCallOnNullableType fail on `requireNotNull`/`checkNotNull`

### ISSUE THREAD

**Issue by yogurtearl** [detekt/detekt#3293]
`UnsafeCallOnNullableType` detekts use of `!!`, but devs like to "workaround" this by using `requireNotNull`/`checkNotNull`

The single param version of these functions are essentially the same as `!!`, except `!!` throws NPE, `requireNotNull` throws `IllegalArgumentException`, `checkNotNull` throws `IllegalStateException`, but swapping out `!!` for 
`requireNotNull`/`checkNotNull` to avoid the detekt warning does not really solve anything. 

The 2 param versions of `requireNotNull`/`checkNotNull` that take a `lazyMessage` might add some value if the error is useful, so not sure if detekt should warn on that one (maybe make it optional?)

instead this rule should push people to actually handle the case where it is null, should be handled with `?.` or `?:` 

## Expected Behavior of the rule

Issue a warning for single param usage of `requireNotNull`/`checkNotNull` just like `!!`
Add an option to also warn on usage of the 2 param version? 

## Context

Prevent lots of changes from `!!` to `requireNotNull`/`checkNotNull` just to avoid detekt warnings. :)

**Comment by cortinico** [comment]
I would suggest we keep the `UnsafeCallOnNullableType` inspection strictly related to unsafe calls (`!!`), to avoid creating confusion.

Please note that we also have a `UseCheckNotNull` and a `UseRequireNotNull` that checks for potential scenarios where `requireNotNull`/`checkNotNull` could be useful. 
I think we want to avoid scenarios where we chain detections: a `UseRequireNotNull` gets reported, code is updated and then an `UnsafeCallOnNullableType` gets reported.

Potentially we could create a new rule that detects usages of `requireNotNull`/`checkNotNull` without a meaningful message. I'm unsure of the effectiveness of such rule.

**Comment by BraisGabin** [comment]
> instead this rule should push people to actually handle the case where it is null, should be handled with `?.` or `?:`

I think that the use of `?.` or `?:` is as smelly as `!!`. So I would not push to them.

But I do agree that the use of `requireNotNull` or `checkNotNull` with a description is better. At least the dev should explain why this should not happen and it's kind of documentation.

I don't know if we should expand `UnsafeCallOnNullableType` or `UseCheckNotNull` and `UseRequireNotNull`... But I agree that we should have a rule to encorage the use of two param with `requireNotNull` and `checkNotNull`.

**Comment by schalkms** [comment]
I think that detekt has rules for all mentioned cases.
I'm not sure why there's a need for a new one.

**Comment by yogurtearl** [comment]
`?.`, `?:` and `if(foo != null)` are all ways of explicitly acknowledging the possibility of `null` and handling it separately. 
If possible, it is best to find a way to handle the null gracefully or make the type non-nullable. 

`!!`, `requireNotNull`, `checkNotNull` all throw exceptions when `null` is encountered. 
There is not much difference in these 3 other than `requireNotNull`\`checkNotNull` provide an optional way to have a better error message.

> I think that detekt has rules for all mentioned cases.

@schalkms  What detekt rule can be used to trigger an error on usage of `requireNotNull`/`checkNotNull`

**Comment by cortinico** [comment]
> @schalkms What detekt rule can be used to trigger an error on usage of `requireNotNull`/`checkNotNull`

If we end up adding a rule for this scenario, it will most likely separate from `UnsafeCallOnNullableType`. 

Having such rule (say `RequireNotNullOnNullableType`) will potentially clash with another rule we already have `UseRequireNotNull`.

**Comment by schalkms** [comment]
> What detekt rule can be used to trigger an error on usage of requireNotNull/checkNotNull

- `UnsafeCallOnNullableType`
- `UseCheckNotNull`
- `UseRequireNotNull`

I still think that custom detekt rules are the solution here.

This case can simply be avoided by enabling `UnsafeCallOnNullableType`.

```kotlin
fun f(str: String?) {
    checkNotNull(str)
    println(str!!)
}
```

**Comment by schalkms** [comment]
Thanks for the feedback to all of you guys. I'm closing this due the mentioned arguments.
